### DISCUSSION TITLE
Support for multiple materials

### DISCUSSION THREAD

**Discussion by YangZhao16** [thelfer/MFrontGenericInterfaceSupport#81] [Q&A] [answered]
@thelfer Hi, I am a new user of FeniCS for solid mechanics, and I come across MFront recently. Very impressive work!

I wonder if MFront now supports multiple materials or subdomains for a given problem? In my case, I usually introduce weak elements to initialize localized shear bands, and I wonder if multiple materials are supported/will be supported by MFront.

Looking forward to your reply!

**Comment by thelfer** [comment]
@YangZhao16 Thanks for you interest.

Per se, multiple materials is not an issue related to `MFront`. The trouble is more on the `FEniCS` side (on the legacy branch, not `FEniCS-X`). We plan to work on this with @bleyerj when we will port `mgis.fenics` to `FEniCS-X`.

For the moment, your trouble seems to be solvable by using non uniform material properties, a feature which is already handled by `mgis.fenics` (or by the `python` bindings of `mgis` if you don't use `mgis.fenics`, which is a shame :)), is it ?

**Reply by YangZhao16** [reply]
Hi @thelfer,

Thank you for your reply!

I've been poking around with the 2D vonMise plasticity example, and I got stuck when I was trying to divide the mesh into two subdomains with different material parameters. I am glad to hear that it is solvable now with mgis by using non uniform material properties. Could you please point me to the documentation, example, or specific keywords that I can use to learn how to do that?

I really appreciate your assistance!
Yang.

**Reply by bleyerj** [reply]
Hi,
I don't think that we have a current documented demo which uses non uniform material properties (yet). But you can simply pass in the material parameters dictionary a dolfin `Expression` or a `Function` defined on some suitable function space (DG0 for instance for piecewise constant material properties). For instance:
```
E1 = 200.
E2 = 400.
E3 = 200.
n12 = 0.3
n13 = 0.3
n23 = 0.3
m12 = 100.
m13 = 100.
m23 = 100.
V0 = FunctionSpace(mesh, "DG", 0)
E3_f = Function(V0)
E3_f.vector()[:] = E3
mat_prop = {"YoungModulus1": Constant(E1),
            "YoungModulus2": Expression("E", E=E2, degree=0),
            "YoungModulus3": E3_f,
            "PoissonRatio12": n12,
            "PoissonRatio13": n13,
            "PoissonRatio23": n23,
            "ShearModulus12": m12,
            "ShearModulus13": m13,
            "ShearModulus23": m23}
material = mf.MFrontNonlinearMaterial("./src/libBehaviour.so",
                                      "OrthotropicElastic", hypothesis="3d",
                                      material_properties=mat_prop,

**Reply by thelfer** [reply]
Thanks @bleyerj !

**Comment by YangZhao16** [comment, marked as answer]
Thank you @bleyerj and @thelfer for the help! :)
